File: iphone/Classes/FBConnect/FBStreamDialog.m
/*
 * Copyright 2009 Facebook
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FBStreamDialog.h"
#import "FBSession.h"

///////////////////////////////////////////////////////////////////////////////////////////////////
// global

static NSString* kStreamURL = @"http://www.facebook.com/connect/prompt_feed.php";

///////////////////////////////////////////////////////////////////////////////////////////////////

@implementation FBStreamDialog

@synthesize attachment        = _attachment,
		    actionLinks       = _actionLinks,
            targetId          = _targetId,
            userMessagePrompt = _userMessagePrompt;

///////////////////////////////////////////////////////////////////////////////////////////////////
// NSObject

- (id)initWithSession:(FBSession*)session {
	if (self = [super initWithSession:session]) {
		_attachment        = @"";
		_actionLinks       = @"";
		_targetId          = @"";
		_userMessagePrompt = @"";
	}
	return self;
}

- (void)dealloc {
	[_attachment        release];
	[_actionLinks       release];
	[_targetId          release];
	[_userMessagePrompt release];
	[super dealloc];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
// FBDialog

- (void)load {
	NSDictionary* getParams = [NSDictionary dictionaryWithObjectsAndKeys:
							   @"touch", @"display", nil];
	
	NSDictionary* postParams = [NSDictionary dictionaryWithObjectsAndKeys:
								_session.apiKey,      @"api_key",
								_session.sessionKey,  @"session_key",
								@"1",                 @"preview",
								@"fbconnect:success", @"callback",
								@"fbconnect:cancel",  @"cancel",
								_attachment,          @"attachment",
								_actionLinks,         @"action_links",
								_targetId,            @"target_id",
								_userMessagePrompt,   @"user_message_prompt",
								nil];
	
	[self loadURL:kStreamURL method:@"POST" get:getParams post:postParams];
}

@end

File: android/modules/facebook/src/ti/modules/titanium/facebook/FBStreamDialog.java
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2010 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package ti.modules.titanium.facebook;

import java.util.HashMap;
import java.util.Map;

import org.appcelerator.titanium.util.Log;

import android.app.Activity;

/**
 * Dialog for handling stream posts
 * 
 */
public class FBStreamDialog extends FBDialog {
	private static final String LOG = FBPermissionDialog.class.getSimpleName();
	private static final String FB_STREAM_URL = "http://www.facebook.com/connect/prompt_feed.php";

	private final String attachment;
	private final String actionLinks;
	private final String targetId;
	private final String userMessagePrompt;

	/**
	 * @param context
	 * @param session
	 */
	public FBStreamDialog(Activity context, FBSession session,
			FacebookModule tb, String attachment, String actionLinks,
			String targetId, String userMessagePrompt) {
		super(context, session);
		this.attachment = attachment;
		this.actionLinks = actionLinks;
		this.targetId = targetId;
		this.userMessagePrompt = userMessagePrompt;
	}

	@Override
	protected void load() {
		Map<String, String> params = new HashMap<String, String>(1);
		params.put("display", "touch");
		params.put("callback", "fbconnect://success");
		params.put("cancel", "fbconnect:cancel");
		// See http://github.com/facebook/facebook-android-sdk/blob/ef53183c59dac75cda7b6787a7fcb6f4bcd35b7b/facebook/src/com/facebook/android/Facebook.java#L42
		// for those urls.
		try {
			Map<String, String> postParams = new HashMap<String, String>(8);
			postParams.put("api_key", mSession.getApiKey());
			if (FacebookModule.usingOauth) {
				postParams.put("access_token", mSession.getAccessToken());
			} else {
				postParams.put("session_key", mSession.getSessionKey());
			}
			postParams.put("preview", "1");
			postParams.put("attachment", attachment);
			postParams.put("action_links", actionLinks);
			postParams.put("target_id", targetId);
			postParams.put("user_message_prompt", userMessagePrompt);

			this.loadURL(FB_STREAM_URL, "post", params, postParams);
		} catch (Exception e) {
			Log.e(LOG, "Error loading URL: " + FB_STREAM_URL, e);
		}
	}

}
